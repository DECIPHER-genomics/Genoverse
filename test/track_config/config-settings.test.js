const { Genoverse, afterTest } = require('../utils');

describe('Config settings', () => {
  afterEach(afterTest);

  const trackDef = {
    data: [
      { chr: 1, start: 1,  end: 10, type: 'A', subtype: 1, label: 'A1', color: 'black'  },
      { chr: 1, start: 21, end: 30, type: 'A', subtype: 2, label: 'A2', color: 'blue'   },
      { chr: 1, start: 41, end: 50, type: 'B', subtype: 1, label: 'B1', color: 'orange' },
      { chr: 1, start: 61, end: 70, type: 'B', subtype: 2, label: 'B2', color: 'purple' },
    ],
    configSettings: {
      type: {
        all : { featureFilter: false },
        a   : { featureFilter: feature => feature.type === 'A' },
        b   : { featureFilter: feature => feature.type === 'B' },
      },
      subtype: {
        all : { featureFilter: false },
        1   : { featureFilter: feature => feature.subtype === 1 },
        2   : { featureFilter: feature => feature.subtype === 2 },
      },
      colorscheme: {
        default : { beforeDrawFeature: false },
        red     : { beforeDrawFeature: (f) => { f.color = 'red';   } },
        green   : { beforeDrawFeature: (f) => { f.color = 'green'; } },
      },
      squish: {
        true: {
          featureHeight : 2,
          featureMargin : { top: 1, right: 1, bottom: 1, left: 0 },
          labels        : false,
        },
        false: {
          featureHeight : 6,
          featureMargin : { top: 2, right: 2, bottom: 2, left: 0 },
          labels        : true,
        },
      },
    },
    defaultConfig: {
      type        : 'all',
      subtype     : 'all',
      colorscheme : 'default',
      squish      : false,
    },
  };

  describe('Track attributes are set correctly', () => {
    const [ track ] = new Genoverse({
      chr            : 1,
      start          : 1,
      end            : 100,
      chromosomeSize : 100,
      tracks         : [ Genoverse.Track.extend($.extend(true, {}, trackDef)) ],
    }).tracks;

    const checkSettings = ($track, confType) => {
      const conf = $track[confType] || $track.config;

      Object.keys(conf).forEach(
        (type) => {
          const settings = $track.configSettings[type][conf[type]];

          Object.entries(settings).forEach(
            ([ key, value ]) => {
              if (key === 'featureFilter') {
                expect(
                  $track.prop('featureFilters').filter(f => f === value).length
                ).toBe(
                  value === false ? 0 : 1
                );
              } else {
                expect($track.prop(key)).toEqual(value);
              }
            }
          );
        }
      );
    };

    it('Initial settings are determined by defaultConfig', () => { checkSettings(track, 'defaultConfig'); });

    describe('Changing config settings', () => {
      Object.entries(track.configSettings).forEach(
        ([ type, configSetting ]) => {
          Object.keys(configSetting).forEach(
            (config) => {
              it(`${type} = ${config}`, () => {
                track.setConfig(type, config);
                expect(track.config[type]).toBe(config);
                checkSettings(track);
              });
            }
          );
        }
      );
    });
  });

  describe('Track, model, view and controller can all have properties and functions set by config', () => {
    let testNow = false;

    const test = (obj, a, b) => {
      if (testNow) {
        expect(a).toEqual(b);
      }
    };

    const [ track ] = new Genoverse({
      chr            : 1,
      start          : 1,
      end            : 100,
      chromosomeSize : 100,
      tracks         : [
        Genoverse.Track.extend({
          afterDraw: function () {
            const d = this.prop('_testDeferred');

            if (d) {
              d.resolve();
            }
          },
          data           : [],
          name           : 'test',
          configSettings : {
            test: {
              1: {
                name           : 1,                      // track property
                clickTolerance : 1,                      // controller property
                dataBuffer     : { start: 100, end: 0 }, // model property that gets used by setDefaults to set dataBufferStart
                dataType       : 1,                      // model property that isn't touched by setDefaults
                featureMargin  : { top: 1 },             // view property that gets modified by setDefaults
                minScaledWidth : 1,                      // view property that isn't touched by setDefaults
                setMVC         : function (...args) {
                  const rtn = this.base(...args);

                  test('track', this.name, 'test - 1');

                  return rtn;
                },
                render: function (...args) {
                  test('controller', this.clickTolerance,  1);

                  return this.base(...args);
                },
                findFeatures: function (...args) {
                  test('model', this.dataBufferStart, 100);
                  test('model', this.dataType,        1);

                  return this.base(...args);
                },
                draw: function (...args) {
                  test('view', this.featureMargin,  { top: 1, right: 0, bottom: 0, left: 0 });
                  test('view', this.minScaledWidth, 1);

                  return this.base(...args);
                },
                _testDeferred: $.Deferred(),
              },
              2: {
                name           : 2,
                clickTolerance : 2,
                dataBuffer     : { start: 200, end: 100 },
                dataType       : 2,
                featureMargin  : { right: 1, left: 2 },
                minScaledWidth : 2,
                setMVC         : function (...args) {
                  const rtn = this.base(...args);

                  test('track', this.name, 'test - 2');

                  return rtn;
                },
                render: function (...args) {
                  test('controller', this.clickTolerance, 2);

                  return this.base(...args);
                },
                findFeatures: function (...args) {
                  test('model', this.dataBuffer, { start: 200, end: 100 });
                  test('model', this.dataType,   2);

                  return this.base(...args);
                },
                draw: function (...args) {
                  test('view',  this.featureMargin, { top: 0, right: 1, bottom: 0, left: 2 });
                  test('view', this.minScaledWidth, 2);

                  return this.base(...args);
                },
                _testDeferred: $.Deferred(),
              },
            },
          },
        }),
      ],
    }).tracks;

    it('config 1', () => {
      testNow = true;
      track.setConfig('test', 1);

      return track._testDeferred;
    });

    it('config 2', () => {
      track.setConfig('test', 2);

      return track._testDeferred;
    });
  });

  describe('Config and zoom settings interact correctly', () => {
    const genoverse = new Genoverse({
      chr            : 1,
      start          : 1,
      end            : 20,
      chromosomeSize : 9e99,
      tracks         : [
        Genoverse.Track.extend({
          data           : [],
          url            : false,
          legend         : false,
          configSettings : {
            test: {
              1 : { xhrFields: { test: 1 }, depth: 1 },
              2 : { xhrFields: { test: 2 }, depth: 2 },
              3 : { xhrFields: { test: 3 }, depth: 3 },
            },
          },
          defaultConfig : { test: 1 },
          1             : { view: Genoverse.Track.View.extend({ _testId: 1    }), model: Genoverse.Track.Model.extend({ _testId: 1    }) },
          1000          : { view: Genoverse.Track.View.extend({ _testId: 1000 }), model: Genoverse.Track.Model.extend({ _testId: 1000 }) },
        }),
      ],
    });

    const [ track ] = genoverse.tracks;

    it('Initial properties are correct', () => {
      expect(track.prop('xhrFields')).toEqual({ test: 1 });
      expect(track.prop('depth')).toBe(1);

      expect(track.model._testId).toBe(1);
      expect(track.view._testId).toBe(1);
    });

    it('After setConfig and then zooming out to a new model/view, properties defined in setConfig are still used', () => {
      track.setConfig('test', 2);
      genoverse.moveTo(1, 1, 1001);

      expect(track.prop('xhrFields')).toEqual({ test: 2 });
      expect(track.prop('depth')).toBe(2);

      expect(track.model._testId).toBe(1000);
      expect(track.view._testId).toBe(1000);
    });

    it('After zooming back in to an existing model/view, properties defined in previous setConfig are still used', () => {
      genoverse.moveTo(1, 1, 20);

      expect(track.prop('xhrFields')).toEqual({ test: 2 });
      expect(track.prop('depth')).toBe(2);

      expect(track.model._testId).toBe(1);
      expect(track.view._testId).toBe(1);
    });

    it('After setConfig and then zooming out to an existing model/view, properties defined in setConfig are still used', () => {
      track.setConfig('test', 3);
      genoverse.moveTo(1, 1, 1001);

      expect(track.prop('xhrFields')).toEqual({ test: 3 });
      expect(track.prop('depth')).toBe(3);

      expect(track.model._testId).toBe(1000);
      expect(track.view._testId).toBe(1000);
    });
  });

  describe('Features are drawn correctly', () => {
    const [ track ] = new Genoverse({
      chr            : 1,
      start          : 1,
      end            : 100,
      width          : 100,
      chromosomeSize : 1000,
      isStatic       : true,
      tracks         : [ Genoverse.Track.extend(trackDef) ],
    }).tracks;

    [
      { type: 'a',   subtype: 'all' },
      { type: 'b',   subtype: 'all' },
      { type: 'all', subtype: 1     },
      { type: 'a',   subtype: 1     },
      { type: 'a',   subtype: 2     },
      { type: 'b',   subtype: 1     },
      { type: 'b',   subtype: 2     },

      { squish: true                        },
      { squish: true,  colorscheme: 'red'   },
      { squish: false, colorscheme: 'green' },

      { type: 'all', subtype: 'all', colorscheme: 'default' },
    ].forEach(
      (config) => {
        it(`when ${Object.entries(config).map(([ k, v ]) => `${k} = ${v}`).join(', ')}`, (done) => {
          track.afterRender = function (f, image) {
            expect(image[0].src).toMatchSnapshot();
            done();
          };

          track.setConfig(config);
        });
      }
    );
  });
});
