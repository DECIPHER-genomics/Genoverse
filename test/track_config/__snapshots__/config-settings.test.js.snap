// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config settings Features are drawn correctly when squish = false, colorscheme = green 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABKUlEQVRoge3XvStFYRwH8M+9rnQnYfM2sMhqsjL6T9hMBvEkxUTKzCQyYTKYMPgLZJHBxGAQ6UoYvFwlZdA5N+f3Gc/z1PPtfHtOv0MI4WelvANkLnn55b5c3k05j0PDz6KQBhOFNJgopMFEISGEEMI/UcQ/9QoecfH+5F7ZhBlH5vV4so4B1LAsWc0yXlGnrJqkX9KPGc8Wv6zNSboxjCnJYJbBilpIXUkVN2DapeQQJNc4Q2eWcSpZHtZAWiTnaPaiFaPfdrzdjD6cZBmsqDfk45PVizFs2Nb0ubqgDZsYl9xlGayohdQlx+hwqgssqarZwYpkP+s4UcicITRrdyWpuLWlZF+ylkecoo+9ZTx4m6b2JCM4UB+JKZk0azePqCGEEEII4a+8Ajr+OQqmnqEWAAAAAElFTkSuQmCC"`;

exports[`Config settings Features are drawn correctly when squish = true 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAOCAYAAADZjbloAAAABmJLR0QA/wD/AP+gvaeTAAAAMUlEQVRYhe3QsQ0AIBACQHTy31xXMBa+xV1NAiHhK6N7wGuVWoe5lm9mRykAAAAAXNhqIQMEUroGNQAAAABJRU5ErkJggg=="`;

exports[`Config settings Features are drawn correctly when squish = true, colorscheme = red 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAOCAYAAADZjbloAAAABmJLR0QA/wD/AP+gvaeTAAAAMklEQVRYhe3QsQ0AIAwDsJT/fy4vIAbKYM+REiXhKzU94LVO+iRXQ9+siVIAAAAAuLABJtwCBO+3j10AAAAASUVORK5CYII="`;

exports[`Config settings Features are drawn correctly when type = a, subtype = 1 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAAA3UlEQVRoge3XIU4DQRQG4I/6Kg5ACCXBYzkAwVCBqeQAnAeDIpg9ARLDDXAQaCCpqidBgdht0hYQJDC7Yf/PzGR2xZ+8vLezRMT3NvD+g3fjjw3aDhCrUpCOSUE6JgXpmBQkIiLiHxqr/9r31s6vMMdd8UQ9tHzLmuC2WZdd4KhYogBDzLCL+y+ej6RDilh0yDGu8aAeT/utJeq5RUEmqJp95fPYioI28YopHvGMF6vfl4ysQgY4wSW2sYMtPOGgxVy9doPDtbMznDf7St0xb+oOOi0XLSIiIiLiN3wASXQc9dTYvxkAAAAASUVORK5CYII="`;

exports[`Config settings Features are drawn correctly when type = a, subtype = 2 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABDklEQVRoge3XLUsEURiG4Xs1CRYRbFsMIlvtBlGwrb/BYnG71W7Xv2CwWYQFFYvVJoKIotFq9DbsiOsufgR3Z2SeCwYO7xzmPPBy5nAgIr7W+PtP6i+XHsHa/99E2QHiszSkYtKQiklDKiYNiYiIiJ+40bu1u9hXa4Jd8BG8BbfLy1c7HoIX4G5frQkuF+O5ojGtcvLVitPgE7gAXn8zrwuuji9X9Y3qHtIGTqBxAzyDS8NTbAHzwOWIMsQHj8G1YtwB9wbez4BX4Pr4s9WOs+ALeFcc3PfgA1jsRqfAc3Cz3Jy14RZ4MFA76x3mToJH4E452WrJ0+FfkR1wH1wBX4ud8/60y8kZEREREfXxBqGPXOs0NnobAAAAAElFTkSuQmCC"`;

exports[`Config settings Features are drawn correctly when type = a, subtype = all 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABgUlEQVRoge3XMWsUQRgG4OdMJWgRAnZpREXSim0QsbC7tFdqExDT2/orkh8gmCtSixBIJI29jURCSDCV2EkgjZ/FzsFljcdtNuCx+R5YdnZ2eXeWj5llSCn9Ww/R4NkpxJR5vSnzrpcb/3sA6bwsyIzJgsyYLMiMyYKklFJKHbSi2rU/rPW/ww98aRYXK9WuPcbyYpHYJr4TB8TrViPuuCH28LbW/wSPNS/IkNgjxvJikVgu7TulMEuXHnGH3cYJ7mP/gvv3NCpI3CJOiAfE1wnPbRPPGo2040b7kD4+4JtqeXrUMrePj/T28ZO4IC+WcBefW76rU0YFGaiWLOU8aJk7wPvS3vw7L+bL/Vf0frV8V+cs4BSHOMARjp3fxTdYsmKBOCUOy4/7iDgmSl7cJD4RL6/uE7plFRu1vl0sj103KcgqUcuL3epnHnPEFvHm0qO9BnbwvNa3hvXSHqpmzJlqBr2YHBc7RC0v1oh14inxu8yc0dFv/wkppZRSSilN8gcDVXYqnpp7DQAAAABJRU5ErkJggg=="`;

exports[`Config settings Features are drawn correctly when type = all, subtype = 1 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABiElEQVRoge3Xv2vUYBjA8c8brxQHcaiboMhBQcGpq4s4CHXQwcH2uEFXd0dn/4hO2ggniG5utWBXd0HOgoNUBJ0E63CvQ3P2tBUV75Igz2dJ8ibDFx7yixDCryXkv7i2drn8s77Ua6Zv2oqmA8KPYiAtEwNpmRhIy8RAQgghhP/E5N/tNTzBWbyaWF/HZezgfH1pe/JzHe98xXa19NnI7dT3AnJZ9SU7abX+vmmb/MpawVa1nbSG5dqKDreberqppyu5K7n3/czImtx439SMB3IMF3DLwYFs4lOdUb9xVPJxfJD6NuVW9f2TTrW9imd4jQ9Ywsumog4xn0tDzMmOK1xqOmhWxnfICgbV/sDBu6Rp40fWKckVI2V+5EjTUbPQwQIu4hxG1VrCneq4VdKqrVxa8MVJvG26Z9oKXMd9nEEXp/HG3juldXJpCXNOeN90yywUuIGnP60/tv/YGmADixjiZm11++ZzaZhL27IHRvpp2S7kdQOFDdliLg3zw0b6QgghhBBCq30D6/ZWOtSjL1sAAAAASUVORK5CYII="`;

exports[`Config settings Features are drawn correctly when type = all, subtype = all, colorscheme = default 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAAC2ElEQVRoge3XP4gcVQCA8d/b3BEOlGBOEIyJxG30wOqqgI2xEEwRC1Hujmu00mATLVKEy+M4SGwiASsJCpLF3FkY0ySCkJBcYWMbRDktIpIoKgQlroR9Fju7mbtjz9k/4Cnva2aYWb79dt+8mTdkMpneBKQ+PluBVNEXKvlSo1pfmKvWF8VKvihW/L2jpfZvfGmmN3lAthl5QLYZeUC2GXlAMplMJpP5n1B+G30Rn+IpfF06fg7P4xaerq5OJV8ofGkvPsSTaOJdwntbWq4Y86O/8H1x6A8tR8K865AaRV9wK8z+c18Ux1jvq6kdWbBwfcnS3nvureuL4pZ9o6a8yprBarEtcxYvDODu5VskPIYDOEaaquBqhjn1MKcuWBCc6p5pOSv13deMYj2KdSy0tE6Vzi1GsdsXxSp9I6MzIA/iGbxq8x94Fb/1p00PFL7X8Mr94+Em4Vqx/5P2THy0z+YJwa9d47yrUr99pSJhgrbvuOM3o3gNojho31CMFdvDuIRv8TOm8dUQ3sP4nPAN6RfSNGGDL03hCXxZwbczNaxhXLJLzXNDtMHOKK5hPEm72OwrZkbVvpHRmSEzWC72l22eJf0yg4+L/fObfemh4vwbhN8r+Dq3rH2CQ1oaacWOIfo6t6x9OITGipWu76ST3b4oVukbGTVM4lm8jzW8hZcN/BafJnGw7UtreLvtS4UvTeACzhAu92sPs1Yx6U97ButbTxRXMXnDjT1w2umJpuYFnIli333DUsNL+Aj7Ucfj+E77GTAIhS/sJ9QJJV/aob1qu0T4YBB5apjGuIfdHrBvHYsWpzG+2+7bURy74875IFyO4kB9wxJwBe+gfDW8iSm8rn0LO4BH8AOWtJeuPUiFr3z1p47vE3zh/pITjhI+28JWXvbWJHclx8K8i5DOWRZKfcFSmO3dt2HZW8Nd7dXUxSge3NgXhKMnnOjZl8lkMplMJpP5L/A3+PfT6KCa/bcAAAAASUVORK5CYII="`;

exports[`Config settings Features are drawn correctly when type = b, subtype = 1 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABFElEQVRoge3XsUrDUBiG4fePLdLJQcFBcOns5OTsINQ7SMgVeA9ODl6E4NIz6E1U0NsobqKLTqJ1yO9iiw5ihoRzjnzP1qTDGz6StCAiv7PYAX/xgLf5nlXpX0sbRewA+UmDJEaDJEaDJEaDiIiI/BPJ/7v1GQMe+ADuvw690nBiNbcAHpgCRxiPVrIXLbQjufzKWljF2CrGGKcY56szDRc4k4htncplkO9GGM/LD1Zzg/MSM6hLg9gBLa17YA4McTYoOIwd1Jdc7pDlI2sX45iG4NesxY7qQy6DrFjJHbDJOzuxW/qQ3SAe2AeGbPEUu6UPub1DCpw3nNomLAB8yhXGAc62B+YYZ1ZyGblXRERERJLyCTOxNkzPojUJAAAAAElFTkSuQmCC"`;

exports[`Config settings Features are drawn correctly when type = b, subtype = 2 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAABRUlEQVRoge3Xr0tDURjG8e+5bsiSOJu/gitiNVk1+p9oMxnG9jAGM00EsyZRltS0YFKDf8GwiMGkwTAUuTI2g9sUYWCQcy+e9xPvOXAe7sO5vBeMMaO5pAP4JtT75b5E3k2UxKFmNCskZayQlLFCUsYKMcYYY/6JEP/UM8A7cN9/9BoRbZYoXVWpznXoHAKLQAzsCu37zBfqlBULFYQKQKlLd+fbWkVoFlgBtoWWfAYLtZAhh8sBzwBFig9ClwBCT8AtMO0zT8bnYSkyLnQHZHv0JoC1nxv6N2MBuPEZLNQbMvhkzQPrwFGDxthgsUZtEjgGNoRefAYLtZAhoWtgqkVrBqBOPRcTnwJ7Qk3feYIvpEJlGcjmyT8KZdq0TxyuKXSQRJ7Qx94IeONzmjoXWgUu+BqJcbitMuWzRMIaY4wxxpg/8gEQ8k4ZTEcv+AAAAABJRU5ErkJggg=="`;

exports[`Config settings Features are drawn correctly when type = b, subtype = all 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAaCAYAAABByvnlAAAABmJLR0QA/wD/AP+gvaeTAAAB5UlEQVRoge3Xv0uVURgH8M+5KeEkaRBkOeQSQZOTaw1B/QeK/0BuTQ6iDyLoZASNkYuXyqmcHIIiHfoLpEUaisiGAinMkPs2+CMrzHu7N/XG+Wzve16+55z34bw8L1mW7S8d9QIOUpQV1TyXBqrbS4iq8kIcybspHcWk2f5yQY6ZXJBjJhfkmMkFybIsy7L/xPH/U3+mxTvf8Hr71hcVQ2nQIhRls7gmeZ/6XT4oL0QLP+eVlIZGjS5OmDi/aXMGF7GB2yHuNnxTf9AsXdZGGtCTBvRIRiVTuyMV9xSu15oXoidED0YrKlN7xsZDnEMfhkNcqn/51WuWguzVJvm4c5EGPVf49LdhSWpjK2/EyJsQLyDEB7zC2TrXW5OWw5ysDieLshW0KrQruVpvXogVtBaKdn7P2z4ZF/Cyzrlq0iwnZOeT1S25oaJczDlRT972J6sbN1CeM7ebN2nyFB7gZojPda69Js1SkF2p3xI6fdXViLwQS+hcttwF06bbNmw8xp0QC42YoxZNV5CirBetTlttRN648V60duhYDdGyZu1hkhZC3G9Efq2are0tKawrDKdB81DMeiTpwxm8lUykfjP75f3S9pawbqubmg9xBU/9aIkl6daYsSf/bINZlmVZlmXZIfgOYVyDipGMt8MAAAAASUVORK5CYII="`;
